left associative scalar math support

philwalk opened the pull request:
Left add / subtract / multiply of matrices.
Already had implemented right add / subtract multiply as methods on `Mat` class.
This adds extensions to types `Double` and `Int` to that the following two expressions have the same semantics:
```scala
val m = Mat[1,2](1,2)
val mat1 = 2 * m
val mat2 = m * 2
```

tests included

dragonfly-ai commented on slash/shared/src/main/scala/slash/matrix/Mat.scala:
```diff
@@ -795,14 +794,67 @@ class Mat[M <: Int, N <: Int](val values: NArray[Double])(using ValueOf[M], Valu
+    }
+    this
+  }
+
```
Both implementations of `addScalar` use nested for loops where a single while loop will do.  Instead, let's do this:
```scala
def addScalar(d: Double)(using ValueOf[N]): Mat[M,N] = {
  var i:Int = 0
  while (i < values.length){
    values(i) += d
    i = i + 1
  }
  this
}

inline def addScalar(n: Int)(using ValueOf[N]): Mat[M,N] = addScalar(n.toDouble)
```

dragonfly-ai commented on slash/shared/src/main/scala/slash/matrix/Mat.scala:
```diff
@@ -795,14 +794,67 @@ class Mat[M <: Int, N <: Int](val values: NArray[Double])(using ValueOf[M], Valu
+  inline def - (s: Int): Mat[M, N] = copy.addScalar(-s)
+
+  inline def -= (s:Double):Mat[M, N] = addScalar(-s)
+  inline def -= (s:Int):Mat[M, N] = addScalar(-s)
```
I still think you don't need the int variants of these methods.  [Scala knows how to coerce Int to Double](https://scastie.scala-lang.org/vnNZ3KNLSsSNcF000OMnUQ).

philwalk replied:
> I still think you don't need the int variants of these methods

agree, will remove.

dragonfly-ai commented:
This can merge if you take out all of the method overloads that take `Int` parameters.  Many remain.

philwalk commented:
I think I got all the overloads taking an `Int`, let me know if I missed anything.

dragonfly-ai commented:
The whole extension in slash/shared/src/main/scala/slash/matrix/package.scala on line 52.

Also, the `-` and `-=` implementations are still incorrect.

philwalk commented:
I removed the unneeded code, added more tests, and all tests now pass.

I left a line that's commented out in matrix/package.scala line 42:

```scala
 inline def *[M <: Int, N <: Int](inline m: Mat[M,N])(using ValueOf[M], ValueOf[N]): Mat[M,N] = m.copy.times(s)
```
I'm suprised that we don't need the extension method on `Double`, because the left multiply in `MatScalaMathTest` line 87   requires `Double` to have the extension method:
```scala
    val matmod = 2 * matrix
```
